Implement simple CopyPropagation based on SSA analysis

cjgillot opened the pull request:
This PR extracts the "copy propagation" logic from https://github.com/rust-lang/rust/pull/106285.

MIR may produce chains of assignment between locals, like `_x = move? _y`.
This PR attempts to remove such chains by unifying locals.

The current implementation is a bit overzealous in turning moves into copies, and in removing storage statements.

cjgillot committed:
Add mir-opt test.

cjgillot committed:
Consider `CopyForDeref` for DestProp.

cjgillot committed:
Make tests unit.

cjgillot committed:
Allow to remove unused definitions without renumbering locals.

cjgillot committed:
Implement SSA CopyProp pass.

cjgillot committed:
Add tests.

cjgillot committed:
Only consider a local to be SSA if assignment dominates all uses.

oli-obk commented:
r? @oli-obk

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit c1a924547eaecff4bc7af3c0de848429a7ac78f1 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk c1a924547eaecff4bc7af3c0de848429a7ac78f1 -->
<!-- homu: {"type":"Approved","sha":"c1a924547eaecff4bc7af3c0de848429a7ac78f1","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

JakobDegen commented on compiler/rustc_mir_transform/src/copy_prop.rs:
```diff
@@ -0,0 +1,286 @@
+            if matches!(body.local_kind(local), LocalKind::Arg) {
+                this.assignments[local] = Set1::One(LocationExtended::Arg);
+            }
+            if borrowed.contains(local) && !decl.ty.is_freeze(tcx, param_env) {
```
The freeze check is not enough here, we should be bailing out on all references. The pass currently changes the bahavior of this code:
```rust
#![feature(custom_mir, core_intrinsics)]
#![allow(unused_assignments)]
extern crate core;
use core::intrinsics::mir::*;

#[custom_mir(dialect = "analysis", phase = "post-cleanup")]
fn f(c: bool) -> bool { mir!({
    let a = c;
    let p = core::ptr::addr_of!(a);
    let p2 = core::ptr::addr_of_mut!(*p);
    *p2 = false;
    RET = c;
    Return()
})}

fn main() {
    assert_eq!(true, f(true));
}
```

This is Miri UB today, but we should not depend on that on stable, especially because there is a real possibility that this may be allowed in the future.

oli-obk commented:
@bors r-

cjgillot committed:
Discard raw pointers from SSA locals.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3953739420/jobs/6770340184) [(plain)](https://github.com/rust-lang/rust/commit/d21e04db51ee2385e301ae17ddc1f43cb39f6b2d/checks/10739275501/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 9511d9dfb1e9
Successfully tagged rust-ci:latest
Built container sha256:9511d9dfb1e96b0e69dcdd678856e65f1b6856183d4c556cf13173e3a729872e
Uploading finished image to https://ci-caches.rust-lang.org/docker/d33e4deadd55ac310bf6d14046967e2ae76df1d0cbcfb3cbdfd57a515110e5c43bea0fc68be40129a2ca62a8c3c0cddee210b82b000006fe1b3d251a4634baf3
upload failed: - to s3://rust-lang-ci-sccache2/docker/d33e4deadd55ac310bf6d14046967e2ae76df1d0cbcfb3cbdfd57a515110e5c43bea0fc68be40129a2ca62a8c3c0cddee210b82b000006fe1b3d251a4634baf3 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
```

</details>

cjgillot committed:
Pacify tidy.

cjgillot committed:
Extract SsaLocals abstraction.

JakobDegen commented:
Sorry, should have thought of this before too. Banning raw pointers isn't enough, references cause problems too. Your pass changes the behavior of this code:

```rust
#![feature(custom_mir, core_intrinsics)]
#![allow(unused_assignments)]
extern crate core;
use core::intrinsics::mir::*;

fn cmp_ref(a: &u8, b: &u8) -> bool {
    std::ptr::eq(a as *const u8, b as *const u8)
}

#[custom_mir(dialect = "analysis", phase = "post-cleanup")]
fn f() -> bool { mir!(
    {
        let a = 5_u8;
        let r1 = &a;
        let b = a;
        let r2 = &b;
        Call(RET, next, cmp_ref(r1, r2))
    }
    next = {
        Return()
    }
)}

fn main() {
    assert!(!f());
}
```

Ideally this would actually be allowed, but we don't currently know how to do that, so need to be conservative in the meantime. See rust-lang/unsafe-code-guidelines#206 for discussion

cjgillot committed:
Do not merge locals that have their address taken.

JakobDegen commented:
Ah, not the change I was expecting, but very nice. Technically this requires a T-opsem FCP these days (since it makes slightly more assumptions about reference immutability than we do already), but that team doesn't really exist yet, so should be fine to land

cjgillot committed:
Use successor location for dominator check.

The assignment is complete only after the statement.
This marks self-assignments `x = x + 1` as non-sSA.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit cbfb98fb8bf1772e059df697c8d15a71b79256d3 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk cbfb98fb8bf1772e059df697c8d15a71b79256d3 -->
<!-- homu: {"type":"Approved","sha":"cbfb98fb8bf1772e059df697c8d15a71b79256d3","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #107309) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 263da251af6b1735487add274a4ed09487c69412 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 263da251af6b1735487add274a4ed09487c69412 -->
<!-- homu: {"type":"Approved","sha":"263da251af6b1735487add274a4ed09487c69412","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 263da251af6b1735487add274a4ed09487c69412 with merge 2a4b00beaa208388c2f140866674605aceab0fe9...
<!-- homu: {"type":"BuildStarted","head_sha":"263da251af6b1735487add274a4ed09487c69412","merge_sha":"2a4b00beaa208388c2f140866674605aceab0fe9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4036896759/jobs/6941137472)
Approved by: oli-obk
Pushing 2a4b00beaa208388c2f140866674605aceab0fe9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4036896759/jobs/6941137472"},"merge_sha":"2a4b00beaa208388c2f140866674605aceab0fe9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4036896759/jobs/6941137472)
Approved by: oli-obk
Pushing 2a4b00beaa208388c2f140866674605aceab0fe9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4036896759/jobs/6941137472"},"merge_sha":"2a4b00beaa208388c2f140866674605aceab0fe9"} -->

rust-timer commented:
Finished benchmarking commit (2a4b00beaa208388c2f140866674605aceab0fe9): [comparison URL](https://perf.rust-lang.org/compare.html?start=d117135f5a9b69ee3adfb1918aa25616758bf692&end=2a4b00beaa208388c2f140866674605aceab0fe9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d117135f5a9b69ee3adfb1918aa25616758bf692&end=2a4b00beaa208388c2f140866674605aceab0fe9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d117135f5a9b69ee3adfb1918aa25616758bf692&end=2a4b00beaa208388c2f140866674605aceab0fe9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.8% | [3.8%, 3.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.8% | [2.8%, 2.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.8% | [3.8%, 3.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d117135f5a9b69ee3adfb1918aa25616758bf692&end=2a4b00beaa208388c2f140866674605aceab0fe9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
